spec/unit/provider_spec.rb

```diff
@@ -436,6 +436,70 @@ def existing_command
       expect(child.specificity).to be > parent.specificity
     end
 
+    ### TODO: A bunch of tests with mulitple defaultfor and notdefaultfor testing specificity
+    ## My intuition: if you don't like jelly beans, and I don't like green jelly beans, I am more likely to eat skittles?
+
+    #Competing default only different in notdefaultfor
+    it "should choose a more specific non-default over a less specific" do
+      skip("Not ready for prime time")
+      Facter.expects(:value).with(:osfamily).at_least_once.returns "redhat"
+      Facter.expects(:value).with(:operatingsystem).at_least_once.returns "centos"
+      Facter.expects(:value).with(:operatingsystemrelease).at_least_once.returns "28"
+
+      one = type.provide(:one) do
+        defaultfor :osfamily => "redhat"
+        notdefaultfor :operatingsystem => "fedora"
+      end
+
+      two = type.provide(:two) do
+        defaultfor :operatingsystem => "redhat"
+        notdefaultfor :operatingsystem => "fedora", :operatingsystemrelease => "27"
+      end
+
+      expect(type.defaultprovider).to eq(two)
```

Comment:
Shouldn't this one default to `one`, as that one says it's the default for `osfamily: 'redhat'`? They're both `notdefaultfor operatingsystem: 'fedora'`, but Facter is stubbed out to report `operatingsystem: 'centos'`, so I wouldn't expect the `notdefaultfor` to come in to play here.

Reply:
I am not sure how to reason about this case: neither one is notdefaultfor, but they are both defaultfor.  The current code would differentiate with inheritance depth but not otherwise.  Does the specificity of the notdefaultfor affect cases where it is not triggered?

Reply:
Talked about this elsewhere, but leaving it here for posterity: If a `defaultfor`/`notdefaultfor` did not match, then the length of its conditions should _not_ be used at all for determining which provider to use.